server/services/utils/functions.js

```diff
@@ -52,6 +53,45 @@ module.exports = ({ strapi }) => {
         });
     },
 
+    buildNestedPaths({items, id = null, field = 'parent', parentPath = null}){
+      return items
+        .filter(entity => {
+          if (entity[field] == null && id === null) {
+            return true;
+          }
+          let data = entity[field];
+          if (data && typeof id === 'string') {
+            data = data.toString();
+          }
+          return (data && data === id) || (isObject(entity[field]) && (entity[field].id === id));
+        })
+        .reduce((acc, entity) => {
+          const path = `${parentPath || ''}/${entity.path}`
+          return [
+            {
+              id: entity.id,
+              parent: parentPath && {
+                id: get(entity, 'parent.id'),
+                path: parentPath,
+              },
+              path
+            },
+            ...this.buildNestedPaths({items, id: entity.id, field, parentPath: path}),
+            ...acc,
+          ];
+        }, [])
+    },
+
+    filterByPath(items, path) {
+      const itemsWithPaths = this.buildNestedPaths({ items }).filter(({path: itemPath}) => itemPath.includes(path));
+      const root = itemsWithPaths.find(({ path: itemPath }) => itemPath === path);
+
+      return {
+        root,
+        items: isNil(root) ? [] : items.filter(({ id }) => (itemsWithPaths.find(v => v.id === id))),
+      }
+    },
+
```

Comment:
@ltsNotMike  can you please write couple of units for those functions?

Reply:
Covered with new test suite